### PR TITLE
[FIXED] MaxMsgPerSubject recovered on corrupt or missing index file.

### PR DESCRIPTION
**Pull request by derekcollison**
On abnormal server exit, for streams or KV with max msgs per subject set we could recover more than N msgs per subject.
This fix allows for recovery of correct state on restart when index files are missing or not current.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/filestore.go [hunk 1]
```diff
@@ -2136,6 +2141,76 @@ func (fs *fileStore) enforceBytesLimit() {
...
+
+	// We want to suppress callbacks from remove during this process
+	// since these should have already been deleted and accounted for.
+	cb := fs.scb
+	fs.scb = nil
+	defer func() { fs.scb = cb }()
+
+	// Also suppress writeIndex info since we will do that manually at the end.
+	fip := fs.fip
+	fs.fip = false
```
Comment (kozlovic) [inline server/filestore.go:2159]: As seen in a test failure, this causes removeMsg() to spin a go routine to do the flush asynchronously, which was detected by a data race. Maybe you need a new boolean in fs struct to handle this case?

**Review by kozlovic — CHANGES_REQUESTED** [review]

File: server/filestore.go (hunk 1 above)
Comment (derekcollison) [reply server/filestore.go:2159]: Ah thanks, its just optimization and we have a flush timer to not go too fast, will remove.

Comment (derekcollison) [reply server/filestore.go]: fixed

File: server/filestore_test.go [hunk 2]
```diff
@@ -4724,3 +4724,77 @@ func TestFileStoreOutOfSpaceRebuildState(t *testing.T) {
...
+	fs.mu.RLock()
+	for _, mb := range fs.blks {
+		mb.mu.Lock()
+		if err := mb.readIndexInfo(); err != nil {
+			mb.mu.Unlock()
+			fs.mu.RUnlock()
+			t.Fatalf("Unexpected error reading index info: %v", err)
+		}
+		if mb.msgs == 0 {
+			mb.mu.Lock()
```
Comment (kozlovic) [inline server/filestore_test.go:4793]: Change to `mb.mu.Unlock()`

**Review by kozlovic — CHANGES_REQUESTED** [review]
Fix needed in the test

File: server/filestore_test.go (hunk 2 above)
Comment (derekcollison) [reply server/filestore_test.go]: Fixed and squashed, sorry about that.

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**